### ISSUE TITLE
if-then-else lines pasted together without whitespace, causes parser failure

### ISSUE THREAD

**Issue by robert-dodier** [rcst/rim#16]
Looks like successive lines of a conditional expression are pasted together without whitespace and sent to Maxima, and then the Maxima parser reports an error. E.g. this example causes an error:
````
$ cat baz.Rmd 
## If-then-else test

```{maxima}
x: 1234;
y: 2345;
```
```{maxima}
if x > y
  then x
  else y;
```
````
The error is:
```
Quitting from lines 8-11 (baz1.Rmd) 
Error in maxima.env$mx$get(code[i]) : 
incorrect syntax: y is not an infix operator
(x > y)then xelse y;
                  ^
```
I can work around the problem by placing parenthesis:
````
$ cat baz1.Rmd 
## If-then-else test

```{maxima}
x: 1234;
y: 2345;
```
```{maxima}
if (x > y)
  then (x)
  else (y);
```
````
and then the output seems to show the pasting-together behavior:
````
$ cat baz1.knit.md 
## If-then-else test


```maxima
(%i5) x: 1234;
```

$$1234$$

```maxima
(%i6) y: 2345;
```

$$2345$$

```maxima
(%i7) if (x > y)then (x)else (y);
```

$$2345$$
````
Not sure what to do. One possibility is to send the lines without pasting them together; I don't know if that interferes with other things.

**Comment by robert-dodier** [comment]
Maybe the following patch is enough? I don't know if `trim` is doing something that we need to keep.
```
$ git diff 
diff --git a/src/utils.cpp b/src/utils.cpp
index 7a40aa5..4a73690 100644
--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -130,9 +130,6 @@ int checkInput(char c, int state)
 // [[Rcpp::export]]
 std::string checkCommand(std::string command)
 { 
-       // remove leading and trailing spaces from command 
-       // alg::trim(command); 
-       trim(command);
        int state = 0; 
        
        // this loop solely checks whether the command issued terminates correctly 
```

**Comment by rcst** [comment]
Thanks! 

The culprit here is actually `collect_ends()` (in R/engine-utils.R), which takes a character vector representing the code chunk where each element contains one line. It pastes together successive lines until the line ends with `;` or `$`. I'll fix this to preserve white space.

I haven't thought about this too deeply and the currently solution is not ideal, since the printed chunk doesn't contain the splitting entered by the user. Maybe a better approach is to just continue sending lines until the line sent ends with `;` or `$`. However, then we need to keep track of where to put the input labels, i.e. the first lines sent in a succession. To do that would mean a bigger changes of `RMaxima$get()`.

The idea was to enable the user to split long lines of code across several lines in a chunk.

**Comment by rcst** [comment]
> Maybe the following patch is enough? I don't know if `trim` is doing something that we need to keep.
> 
> ```
> $ git diff 
> diff --git a/src/utils.cpp b/src/utils.cpp
> index 7a40aa5..4a73690 100644
> --- a/src/utils.cpp
> +++ b/src/utils.cpp
> @@ -130,9 +130,6 @@ int checkInput(char c, int state)
>  // [[Rcpp::export]]
>  std::string checkCommand(std::string command)
>  { 
> -       // remove leading and trailing spaces from command 
> -       // alg::trim(command); 
> -       trim(command);
>         int state = 0; 
>         
>         // this loop solely checks whether the command issued terminates correctly 
> ```

For your given example above, `checkCommand()` is processing the string `"if x > ythen xelse y;"`. 

In addition `checkCommand()` appends a `;` if there is no termination sign.

**Comment by rcst** [comment]
I have implemented a fix for this in `c28da93`

A new internal function `gather()` identifies indices of the lines of code that belong to a single command, these are pasted together including any white spaces in between. This way, code chunks are printed as written by the user, preserving these white space characters. The only exception are empty lines, which are skipped and not printed. I tried it without the skipping of empty lines but this effectively added a line break between the input label and the following code lines.

Do you think that's fine?

I pasted your example into `inst/extdata/test.Rmd` to make it part of the actual unit test.
